<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.31">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Isaac Vock">
<meta name="dcterms.date" content="2025-12-30">

<title>On ‘Fundamental errors in RNA velocity arising from the omission of cell growth’ – isaacvock.github.io</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-e1a5c8363afafaef2c763b6775fbf3ca.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-dfb324f25d9b1687192fa8be62ac8f9c.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">isaacvock.github.io</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../about.html"> 
<span class="menu-text">Research</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../blog.html"> 
<span class="menu-text">Blog</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">On ‘Fundamental errors in RNA velocity arising from the omission of cell growth’</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">nrseq</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Isaac Vock </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">December 30, 2025</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#tldr" id="toc-tldr" class="nav-link" data-scroll-target="#tldr">TLDR</a>
  <ul class="collapse">
  <li><a href="#the-good" id="toc-the-good" class="nav-link" data-scroll-target="#the-good">The Good</a></li>
  <li><a href="#the-bad" id="toc-the-bad" class="nav-link" data-scroll-target="#the-bad">The Bad</a></li>
  <li><a href="#the-ugly" id="toc-the-ugly" class="nav-link" data-scroll-target="#the-ugly">The Ugly</a></li>
  </ul></li>
  <li><a href="#my-criticisms" id="toc-my-criticisms" class="nav-link" data-scroll-target="#my-criticisms">My criticisms</a>
  <ul class="collapse">
  <li><a href="#a-flawed-analysis-paradigm" id="toc-a-flawed-analysis-paradigm" class="nav-link" data-scroll-target="#a-flawed-analysis-paradigm">A flawed analysis paradigm</a></li>
  <li><a href="#flawed-discussions-of-metabolic-labeling-data" id="toc-flawed-discussions-of-metabolic-labeling-data" class="nav-link" data-scroll-target="#flawed-discussions-of-metabolic-labeling-data">Flawed discussions of metabolic labeling data</a></li>
  <li><a href="#ignoring-the-bigger-problems" id="toc-ignoring-the-bigger-problems" class="nav-link" data-scroll-target="#ignoring-the-bigger-problems">Ignoring the bigger problems</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>Recently, <a href="https://www.biorxiv.org/content/10.64898/2025.12.18.695252v1.full">a preprint</a> made some interesting claims about a widespread flaw in the analysis of metabolic labeling RNA-seq/scRNA-seq data. I’ve worked extensively in this space and have written in-depth reviews of metabolic labeling analysis strategies, and thus my interest was piqued. After reading the preprint a couple times, I have a lot of thoughts.</p>
</section>
<section id="tldr" class="level2">
<h2 class="anchored" data-anchor-id="tldr">TLDR</h2>
<p>In summary, there are some things I really like about this work, and even more things that I am critical of.</p>
<section id="the-good" class="level3">
<h3 class="anchored" data-anchor-id="the-good">The Good</h3>
<ul>
<li>This work accurately points out that cellular growth introduces deviations from steady-state in bulk RNA dynamics. This risks biasing standard analyses of RNA degradation kinetics, but as I discuss in <strong>The Bad</strong>, this bias will often be pretty small.</li>
<li>RNA velocity analyses have gotten pretty complicated. There are a litany of ideas of how to take a collection of gene-wise velocity estimates and do things like estimate vector fields, map cell state dynamics, and predict the impacts of perturbations. Instead of tackling all of this complexity head on, the authors focus on a simpler, sometimes neglected, part of the pipeline: actually estimating the gene-wise velocities. I suspect that there are a ton of challenges that need to be intelligently tackled at this first stage, so more work like this is needed.</li>
<li>The authors focus on metabolic labeling-based velocity analyses, which promise to overcome some fundamental weaknesses of splicing-based analyses (e.g., reliance on biased or low intronic coverage, in ability to estimate absolute velocities).</li>
</ul>
</section>
<section id="the-bad" class="level3">
<h3 class="anchored" data-anchor-id="the-bad">The Bad</h3>
<ul>
<li>The authors put forth a flawed analysis workflow that is not accurately reflective of how previous work has formalized estimating velocities with metabolic labeling data. They model absolute molecular content of cells, and assume that UMI counts represent accurate estimates of these quantities.</li>
<li>The authors seem unaware of an existing strategy to address the growth-induced non-steady-state dynamics.</li>
<li>Growth rates assumed throughout the work are fairly high (12 hours is number they often quote). Growth rates differ from context-to-context, but even the fastest growing cell lines that researchers often work with have doubling times on the order of days. Since these typical growth rates are on much longer time scales than RNA degradation (median half-life &lt; 4 hours), as well as the labeling done in these experiments, the impact of growth-induced non-steady-state dynamics are usually going to be quite limited.</li>
<li>Even when growth rates matter though, its affect is global. That is, the growth rate is a global parameter that affects all bulk RNA synthesis kinetics equally. Looking beyond RNA velocity analyses for a minute, many investigations of RNA turnover kinetics using metabolic labeling data only need the ordering of RNA degradation kinetic estimates to be accurate. If there are some biases in the global scale of informed RNA degradation, this can be normalized out when comparing estimates across cell lines, training ML models on degradation rate constant estimates, etc. There are plenty other, much larger sources of such bias (dropout, label concentration ramp up), which can be similarly navigated.</li>
<li>The discussion of several aspects of metabolic labeling experiments is a bit naive or ill-informed.</li>
</ul>
</section>
<section id="the-ugly" class="level3">
<h3 class="anchored" data-anchor-id="the-ugly">The Ugly</h3>
<p>Not necessarily worse than anything covered in <strong>The Bad</strong>, but I couldn’t resist the movie reference.</p>
<p>It’s obvious that the analysis framework but forth by the authors is deeply flawed and not representative of how these data are typically analyzed, in that they seemed to have tried to apply it to their own data only for the facade to crumble:</p>
<blockquote class="blockquote">
<p>We ran a pulse-chase experiment with an additional dye allowing us to identify cells that had not divided during the chase (Methods). … However, the scRNA-Seq results do not display a concordant increase in mRNA (UMI) counts per cell, even after correction for sequencing depth (Figure S5B-D; Methods), including in a repeated experiment. We interpretate [sic] this difference as a technical problem, rather than as an accurate representation of the biology. Because the observed counts are not scaling appropriately with true counts, it is not possible to cleanly estimate the isolated degradation rate, as hoped. More work is needed, therefore, to develop technology or methods to isolate degradation and dilution effects.</p>
</blockquote>
</section>
</section>
<section id="my-criticisms" class="level2">
<h2 class="anchored" data-anchor-id="my-criticisms">My criticisms</h2>
<p>Here I go in to more depth about some of the grievances covered above.</p>
<section id="a-flawed-analysis-paradigm" class="level3">
<h3 class="anchored" data-anchor-id="a-flawed-analysis-paradigm">A flawed analysis paradigm</h3>
<p>The authors put forth an analysis framework that they claim is representative of how metabolic labeling data should be analyzed, at least for the purposes of RNA velocity analyses. In summary:</p>
<ul>
<li>The authors put forth a reasonable and standard model of labeled and unlabeled RNA abundance. The key observation is that the bulk synthesis rate is a function of the number of cells, and this number increases due to cell growth. Formalized:</li>
</ul>
<p><span class="math display">\[
\begin{gather}
\frac{dR}{dt} = k_{syn}(t) - k_{deg} \cdot R \\
k_{syn}(t) = n \cdot k_{syn}^{sc} \\
k_{syn}^{sc} := \text{Average transcription rate in a single-cell}
\end{gather}
\]</span></p>
<ul>
<li>The authors then propose to effectively fit said model to data on absolute single-cell molecular abundances.</li>
</ul>
<p>Step 2 is where some of the challenges arise. First off, how does one even estimate the absolute counts of molecules in a single cell? Given the quote in <strong>The Ugly</strong> in the above section, it seems that the authors assume that UMI counts are the key to this. This is a flawed assumption. If sequenced to saturation (i.e., you sequence every unique molecule), then UMI counts are a measure of the absolute molecular content of <strong>that which made it to the sequencer</strong>. Absolute molecular content that got sequenced is not the same as absolute molecular content of the cells. The authors seem to have learned the hard way that UMI counts can vary for entirely technical reasons that prevent their use as truly quantitative metrics of RNA molecule counts in a cell.</p>
<p>How are metabolic labeling datasets actually analyzed then? I would argue that there are two “gold-standard” analysis strategies:</p>
<ol type="1">
<li>Steady-state NTRs: In this approach, the ratio of labeled (new) to total RNA is analyzed (discussed <a href="https://isaacvock.github.io/posts/nrseq_deepdive/#standard-kinetic-analyses">here</a>). At steady-state, there is a clean relationship between this quantity and turnover kinetics. While a single-cell is rarely at steady-state, the ability to rely on an internally normalized ratio computed on a per-sample basis has a number of statistical advantages, as discussed <a href="https://isaacvock.github.io/posts/nrseq_deepdive/#non-steady-state-modeling">here</a>.</li>
<li>Non-steady-state NTRs: An NTR is calculated, but using two samples (discussed <a href="https://isaacvock.github.io/posts/nrseq_deepdive/#non-steady-state-modeling">here</a>). A sample subjected to metabolic labeling is used to estimate the labeled RNA abundance, and a sample collected at the start of labeling is used to estimate the initial total RNA abundance. RNA abundance estimates come from normalized read counts (e.g., using normalization approaches such as median-of-ratios). Such normalization approaches should be robust in this case, as it is unlikely that there are radical changes in RNA content between the start and end of labeling, especially when using a label time of around 2 hours.</li>
</ol>
<p>The author’s approach is close in essence to 2. But if properly applied, any of the growth rate-induced biases should be eliminated. Thus, I suspect that their choice to think about things on the scale of absolute molecular abundances is leading to the biases they find in their approach when analyzing simulated data.</p>
<p>It is worth noting that many previous studies using metabolic labeling to perform scRNA-seq have relied on something closer to the steady-state NTR approach.</p>
</section>
<section id="flawed-discussions-of-metabolic-labeling-data" class="level3">
<h3 class="anchored" data-anchor-id="flawed-discussions-of-metabolic-labeling-data">Flawed discussions of metabolic labeling data</h3>
</section>
<section id="ignoring-the-bigger-problems" class="level3">
<h3 class="anchored" data-anchor-id="ignoring-the-bigger-problems">Ignoring the bigger problems</h3>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>