<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.554">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Isaac Vock">
<meta name="dcterms.date" content="2024-12-16">

<title>isaacvock.github.io - Introducing NR-seq by simulating NR-seq</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/live-runtime/live-runtime.js" type="module"></script>
<link href="../../site_libs/quarto-contrib/live-runtime/live-runtime.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script type="module" src="../../site_libs/quarto-ojs/quarto-ojs-runtime.js"></script>
<link href="../../site_libs/quarto-ojs/quarto-ojs.css" rel="stylesheet">

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">isaacvock.github.io</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../about.html"> 
<span class="menu-text">Research</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../blog.html"> 
<span class="menu-text">Blog</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Introducing NR-seq by simulating NR-seq</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">nrseq</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Isaac Vock </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">December 16, 2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#a-brief-introduction-to-nr-seq" id="toc-a-brief-introduction-to-nr-seq" class="nav-link active" data-scroll-target="#a-brief-introduction-to-nr-seq">A brief introduction to NR-seq</a></li>
  <li><a href="#understanding-nr-seq-by-simulating-nr-seq" id="toc-understanding-nr-seq-by-simulating-nr-seq" class="nav-link" data-scroll-target="#understanding-nr-seq-by-simulating-nr-seq">Understanding NR-seq by simulating NR-seq</a>
  <ul class="collapse">
  <li><a href="#the-full-shebang" id="toc-the-full-shebang" class="nav-link" data-scroll-target="#the-full-shebang">The full shebang</a></li>
  <li><a href="#mutable-nucleotide-content" id="toc-mutable-nucleotide-content" class="nav-link" data-scroll-target="#mutable-nucleotide-content">Mutable nucleotide content</a></li>
  <li><a href="#mutational-content" id="toc-mutational-content" class="nav-link" data-scroll-target="#mutational-content">Mutational content</a></li>
  <li><a href="#rna-kinetics-and-nr-seq" id="toc-rna-kinetics-and-nr-seq" class="nav-link" data-scroll-target="#rna-kinetics-and-nr-seq">RNA kinetics and NR-seq</a></li>
  <li><a href="#simulating-multiple-transcripts" id="toc-simulating-multiple-transcripts" class="nav-link" data-scroll-target="#simulating-multiple-transcripts">Simulating multiple transcripts</a></li>
  </ul></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary">Summary</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<p>Nucleotide recoding RNA-seq (NR-seq; SLAM-seq, TimeLapse-seq, TUC-seq, etc.), is a set of methods to probe the dynamics of RNA. These methods use metabolic labeling. In this case, metabolic labeling means treating cells with a molecule that looks like a regular nucleotide (e.g., 4-thiouridine, or s4U), which cells incorporate into nascent RNA. NR-seq then employs a chemistry to modify the metabolic label so that a reverse transcriptase identifies it as a nucleotide different from what it originally mimiced (e.g., recoding s4U as a cytosine analog). This allows label incorporation events to be bioinformatically detected as apparent mutations in aligned sequencing reads.</p>
<p>This post introduces the main ideas behind the modeling of NR-seq data.</p>
<section id="a-brief-introduction-to-nr-seq" class="level2">
<h2 class="anchored" data-anchor-id="a-brief-introduction-to-nr-seq">A brief introduction to NR-seq</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="TimeLapse.png" class="img-fluid figure-img"></p>
<figcaption>Schematic of an NR-seq experiment. Adapted from Schofield et al.&nbsp;2018.</figcaption>
</figure>
</div>
<p>Developing a mechanistic understanding of gene expression regulation requires methods to probe the kinetics of RNA synthesis, processing, and degradation. While standard RNA-seq begins to solve this problem, it provides limited information about the kinetics of the processes which determine an RNA’s abundance. Nucleotide recoding RNA-seq (NR-seq; TimeLapse-seq, SLAM-seq, TUC-seq, etc.) overcomes these limitations. NR-seq combines metabolic labeling with novel chemistries that recode the hydrogen bonding pattern of a metabolic label so as to facilitate detection of labeled RNA via these chemically induced mutations in sequencing reads, absolving the need for biochemical enrichment of labeled RNA. By providing information about both overall RNA abundance and the dynamics of nascent and pre-existing RNA, NR-seq resolves the kinetic ambiguities of standard RNA-seq.</p>
<p>One way to intuit how metabolic labeling provides information about the dynamics of RNA is to consider one of the populations being tracked: the old, unlabeled RNA. Since this RNA can only degrade, its dynamics are entirely determined by its turnover kinetics. Combining this with the abundance information provided by standard RNA-seq provides information about the RNA’s synthesis kinetics (transcription + processing).</p>
</section>
<section id="understanding-nr-seq-by-simulating-nr-seq" class="level2">
<h2 class="anchored" data-anchor-id="understanding-nr-seq-by-simulating-nr-seq">Understanding NR-seq by simulating NR-seq</h2>
<p>Here, we will walk through a basic simulation of NR-seq data to give you a sense as to how to think about and interpret NR-seq data.</p>
<section id="the-full-shebang" class="level3">
<h3 class="anchored" data-anchor-id="the-full-shebang">The full shebang</h3>
<p>Below is a summary of the entire simulation which we will walk through piece by piece in the following sections.</p>
<div class="cell">
<div>
<div id="webr-1" class="exercise-cell">

</div>
<script type="webr-1-contents">
eyJjb2RlIjoiIyMjIFNpbXVsYXRpb24gcGFyYW1ldGVyc1xuXG4jIE51bWJlciBvZiB0cmFuc2NyaXB0cyB0byBzaW11bGF0ZVxubnQgPC0gMTAwXG5cbiMgRGlzdHJpYnV0aW9uIHBhcmFtZXRlcnNcbmtkZWdfbG9nbWVhbiA8LSAtMS45XG5rZGVnX2xvZ3NkIDwtIDAuN1xua3N5bl9sb2dtZWFuIDwtIDIuM1xua3N5bl9sb2dzZCA8LSAwLjdcblVjb250X2FscGhhIDwtIDI1XG5VY29udF9iZXRhIDwtIDc1XG5cbiMgRXhwZXJpbWVudGFsIHBhcmFtZXRlcnNcbnNlcWRlcHRoIDwtIDI1MDAwXG5yZWFkbGVuIDwtIDE1MFxudGwgPC0gNFxucG5ldyA8LSAwLjA1XG5wb2xkIDwtIDAuMDAyXG5cblxuIyMjIFNpbXVsYXRlIHRyYW5zY3JpcHQgcGFyYW1ldGVyc1xua2RlZ3MgPC0gcmxub3JtKG50LCBrZGVnX2xvZ21lYW4sIGtkZWdfbG9nc2QpXG5rc3lucyA8LSBybG5vcm0obnQsIGtzeW5fbG9nbWVhbiwga3N5bl9sb2dzZClcblJzcyA8LSBrc3lucyAvIGtkZWdzXG5yZWxfYWJ1bmRhbmNlIDwtIFJzcyAvIHN1bShSc3MpXG5mbnMgPC0gMSAtIGV4cCgta2RlZ3MqdGwpXG5cblVjb250cyA8LSByYmV0YShudCwgVWNvbnRfYWxwaGEsIFVjb250X2JldGEpXG5cbiMjIyBTaW11bGF0ZSByZWFkIGNvdW50cyBmb3IgZWFjaCB0cmFuc2NyaXB0XG5yZWFkc19wZXJfdCA8LSBybXVsdGlub20oMSwgXG4gICAgICAgICAgICAgICAgICAgICAgICAgc2l6ZSA9IHNlcWRlcHRoLFxuICAgICAgICAgICAgICAgICAgICAgICAgIHByb2IgPSByZWxfYWJ1bmRhbmNlKVssMV1cblxuIyMjIFNpbXVsYXRlIHJlYWQtc3BlY2lmaWMgZGF0YVxuXG5uZXduZXNzIDwtIHJiaW5vbShzZXFkZXB0aCxcbiAgICAgICAgICAgICAgICAgIHNpemUgPSAxLFxuICAgICAgICAgICAgICAgICAgcHJvYiA9IHJlcChmbnMsIHRpbWVzID0gcmVhZHNfcGVyX3QpKVxuXG5uVCA8LSByYmlub20oc2VxZGVwdGgsXG4gICAgICAgICAgICAgc2l6ZSA9IHJlYWRsZW4sXG4gICAgICAgICAgICAgcHJvYiA9IHJlcChVY29udHMsIHRpbWVzID0gcmVhZHNfcGVyX3QpKVxuXG5UQyA8LSByYmlub20oc2VxZGVwdGgsXG4gICAgICAgICAgICAgc2l6ZSA9IG5ULFxuICAgICAgICAgICAgIHByb2IgPSBwbmV3ICogbmV3bmVzcyArIHBvbGQpXG5cbiMjIyBDb21waWxlIGRhdGFcblxuc2ltX2RmIDwtIHRpYmJsZShcbiAgdHJhbnNjcmlwdCA9IHBhc3RlMChcInRyYW5zY3JpcHRcIiwgcmVwKDE6bnQsIHRpbWVzID0gcmVhZHNfcGVyX3QpKSxcbiAgVEMgPSBUQyxcbiAgblQgPSBuVCxcbiAgbmV3bmVzcyA9IG5ld25lc3NcbilcblxucGFyX2RmIDwtIHRpYmJsZShcbiAgdHJhbnNjcmlwdCA9IHBhc3RlMChcInRyYW5zY3JpcHRcIiwgMTpudCksXG4gIGtkZWcgPSBrZGVncyxcbiAga3N5biA9IGtzeW5zLFxuICBmbiA9IGZucyxcbiAgcmVhZHMgPSByZWFkc19wZXJfdFxuKVxuXG4jIyMgVmlzdWFsaXplXG4jIFBhbmVsIDE6IGtkZWcgZGlzdHJpYnV0aW9uXG4jIFBhbmVsIDI6IGtzeW4gZGlzdHJpYnV0aW9uXG4jIFBhbmVsIDM6IGZuIGRpc3RyaWJ1dGlvblxuIyBQYW5lbCA0OiByZWFkIGNvdW50IGRpc3RyaWJ1dGlvblxuIyBQYW5lbCA1OiBuZXcgcmVhZCBtdXRhdGlvbiBkaXN0cmlidXRpb25cbiMgUGFuZWwgNjogb2xkIHJlYWQgbXV0YXRpb24gZGlzdHJpYnV0aW9uXG5wMSA8LSBnZ3Bsb3QocGFyX2RmLFxuICAgICAgICAgICAgIGFlcyh4ID0ga2RlZykpICtcbiAgZ2VvbV9oaXN0b2dyYW0oKSArXG4gIHRoZW1lX2NsYXNzaWMoKSArIFxuICB4bGFiKFwia2RlZydzXCIpICsgXG4gIHlsYWIoXCJ0cmFuc2NyaXB0IGNvdW50XCIpXG5cbnAyIDwtIGdncGxvdChwYXJfZGYsXG4gICAgICAgICAgICAgYWVzKHggPSBrc3luKSkgK1xuICBnZW9tX2hpc3RvZ3JhbSgpICtcbiAgdGhlbWVfY2xhc3NpYygpICsgXG4gIHhsYWIoXCJrc3luJ3NcIikgKyBcbiAgeWxhYihcInRyYW5zY3JpcHQgY291bnRcIilcblxucDMgPC0gZ2dwbG90KHBhcl9kZixcbiAgICAgICAgICAgICBhZXMoeCA9IGZucykpICtcbiAgZ2VvbV9oaXN0b2dyYW0oKSArXG4gIHRoZW1lX2NsYXNzaWMoKSArIFxuICB4bGFiKFwiRnJhY3Rpb24gbmV3XCIpICsgXG4gIHlsYWIoXCJ0cmFuc2NyaXB0IGNvdW50XCIpXG5cbnA0IDwtIGdncGxvdChwYXJfZGYsXG4gICAgICAgICAgICAgYWVzKHggPSBsb2cxMChyZWFkcyArIDEpKSkgK1xuICBnZW9tX2hpc3RvZ3JhbSgpICtcbiAgdGhlbWVfY2xhc3NpYygpICsgXG4gIHhsYWIoXCJsb2cxMChyZWFkcyArIDEpXCIpICsgXG4gIHlsYWIoXCJ0cmFuc2NyaXB0IGNvdW50XCIpXG5cbnA1IDwtIGdncGxvdChzaW1fZGYgJT4lIFxuICAgICAgICAgICAgICAgZHBseXI6OmZpbHRlcihuVCA+IDAgJiBuZXduZXNzID09IDApLFxuICAgICAgICAgICAgIGFlcyh4ID0gVEMgLyBuVCkpICtcbiAgZ2VvbV9oaXN0b2dyYW0oZmlsbCA9ICdkYXJrZ3JheScpICtcbiAgdGhlbWVfY2xhc3NpYygpICsgXG4gIHhsYWIoXCJNdXRhdGlvbiByYXRlIGZvciBvbGQgcmVhZHNcIikgKyBcbiAgeWxhYihcInRyYW5zY3JpcHQgY291bnRcIilcblxucDYgPC0gZ2dwbG90KHNpbV9kZiAlPiUgXG4gICAgICAgICAgICAgICBkcGx5cjo6ZmlsdGVyKG5UID4gMCAmIG5ld25lc3MgPT0gMSksXG4gICAgICAgICAgICAgYWVzKHggPSBUQyAvIG5UKSkgK1xuICBnZW9tX2hpc3RvZ3JhbShmaWxsID0gJ2RhcmtyZWQnKSArXG4gIHRoZW1lX2NsYXNzaWMoKSArIFxuICB4bGFiKFwiTXV0YXRpb24gcmF0ZSBmb3IgbmV3IHJlYWRzXCIpICsgXG4gIHlsYWIoXCJ0cmFuc2NyaXB0IGNvdW50XCIpXG5cblxuZ3JpZC5hcnJhbmdlKHAxLCBwMiwgcDMsIHA0LCBwNSwgcDYsXG4gICAgICAgICAgICAgbnJvdyA9IDMsXG4gICAgICAgICAgICAgbmNvbCA9IDIpIiwiYXR0ciI6eyJlZGl0Ijp0cnVlLCJldmFsIjp0cnVlfX0=
</script>
</div>
</div>
</section>
<section id="mutable-nucleotide-content" class="level3">
<h3 class="anchored" data-anchor-id="mutable-nucleotide-content">Mutable nucleotide content</h3>
<p>Different RNAs have different nucleotide content. Sequencing reads in standard short read sequencing experiments typically only sample a small portion of the entire RNA, and thus reads from a given species of RNA will also vary in their nucleotide content. In a standard NR-seq experiments, uridines in an RNA synthesized in the presence of metabolic label have an opportunity to be replaced with s4U. This, we need to simulate the number of Us in each read (prior to nucleotide recoding). Since we typically sequence cDNA, and this is also best thought as the number of Ts in the genomic sequence to which a given read aligned, we will refer to this as the number of T’s.</p>
<p>Let’s consider simulating sequencing reads from a specific RNA species (e.g., a particular transcript isoform) first. In this case, the RNA will have a particular U-content, i.e., the fraction of nucleotides in its exonic sequence that are U’s. This is a simulation parameter, along with the number of reads we want to simulate, and the length of each read. It’s simplest to assume that each nucleotide in a read has some set probability of being a U, and that this probability is the same for all nucleotides. In this case, we can draw the number of U’s (or cDNA T’s) from a <a href="https://en.wikipedia.org/wiki/Binomial_distribution">binomial distribution</a>:</p>
<div class="cell">
<div>
<div id="webr-2" class="exercise-cell">

</div>
<script type="webr-2-contents">
eyJjb2RlIjoiIyBRdWFydG8gbGl2ZSBjYW4gYmUgd2VpcmQgc29tZXRpbWVzXG4jIEJlc3QgdG8gZXhwbGljaXRseSBzZXQgYmFja2dyb3VuZCBhcyB3aGl0ZVxucGFyKGJnID0gXCJ3aGl0ZVwiKVxuXG4jIyMgU2ltdWxhdGlvbiBwYXJhbWV0ZXJzXG5yZWFkcyA8LSAxMDBcblVjb250IDwtIDAuMjVcbnJlYWRsZW4gPC0gMTUwXG5cbiMjIyBTaW11bGF0aW9uXG5uVCA8LSByYmlub20ocmVhZHMsIHNpemUgPSByZWFkbGVuLCBwcm9iID0gVWNvbnQpXG5cblxuIyMjIFZpc3VhbGl6ZVxuaGlzdChuVCkiLCJhdHRyIjp7ImVkaXQiOnRydWUsImV2YWwiOnRydWV9fQ==
</script>
</div>
</div>
</section>
<section id="mutational-content" class="level3">
<h3 class="anchored" data-anchor-id="mutational-content">Mutational content</h3>
<p>The number of mutations in an NR-seq sequencing read provides the information necessary to classify it as having come from either labeled or unlabeled RNA. If a read is from labeled RNA (or RNA that was synthesized during the label time; a given RNA molecule may not incorporate any metabolic label even if it could have), the mutational content of that read is a function of the following things:</p>
<ol type="1">
<li>How many mutable nucleotides are contained in the read, which we simulated above.</li>
<li>How often the metabolic label is incorporated in place of the standard nucleotide. We will also lump into this term the chemical efficiency of recoding, which is usually pretty high (&gt; 80%).</li>
<li>The background mutation rate, due to sequencing/RT/PCR errors, alignment errors, SNPs, etc.</li>
</ol>
<p>We’ve already simulated the first factor. The other two are parameters that we will include in our simulation. Incorporation rates in successful NR-seq experiments are typically around 5%, and background mutation rates can range from nearly 0% to around 0.4%, depending on a number of factors. 0.2% is a fairly typical background mutation rate in my experience.</p>
<p>Finally though, we need to determine the “newness” status of each sequencing read. Each species of RNA will have a characteristic “fraction new”, which is the fraction of RNA molecules present at RNA extraction time that were exposed to metabolic label. We can then model sequencing reads as being randomly drawn from this pool, with fraction new probability of sampling a labeled RNA. These processess are all well modeled with binomial distributions, so simulating all of this looks like:</p>
<div class="cell">
<div>
<div id="webr-3" class="exercise-cell">

</div>
<script type="webr-3-contents">
eyJjb2RlIjoiIyBRdWFydG8gbGl2ZSBjYW4gYmUgd2VpcmQgc29tZXRpbWVzXG4jIEJlc3QgdG8gZXhwbGljaXRseSBzZXQgYmFja2dyb3VuZCBhcyB3aGl0ZVxucGFyKGJnID0gXCJ3aGl0ZVwiKVxuXG4jIyMgU2ltdWxhdGlvbiBwYXJhbWV0ZXJzXG5yZWFkcyA8LSAxMDBcblVjb250IDwtIDAuMjVcbnJlYWRsZW4gPC0gMTUwXG5mbiA8LSAwLjUgIyBmcmFjdGlvbiBuZXdcbnBuZXcgPC0gMC4wNSAjIGluY29ycG9yYXRpb24vcmVjb2RpbmcgcmF0ZVxucG9sZCA8LSAwLjAwMiAjIGJhY2tncm91bmQgbXV0YXRpb24gcmF0ZVxuXG5cbiMjIyBuVCBTaW11bGF0aW9uXG5uVCA8LSByYmlub20ocmVhZHMsIHNpemUgPSByZWFkbGVuLCBwcm9iID0gVWNvbnQpXG5cblxuIyMjIE11dGF0aW9uYWwgY29udGVudCBzaW11bGF0aW9uXG5uZXduZXNzIDwtIHJiaW5vbShyZWFkcywgc2l6ZSA9IDEsIHByb2IgPSBmbilcblRDIDwtIHJiaW5vbShyZWFkcywgc2l6ZSA9IG5ULCBwcm9iID0gcG5ldypuZXduZXNzICsgcG9sZClcblxuIyMjIFZpc3VhbGl6ZVxuaGlzdChUQyAvIG5ULFxuICAgICBtYWluID0gJ0FsbCByZWFkcycsXG4gICAgIGNvbCA9ICdjb3JhbCcpXG5oaXN0KFRDW25ld25lc3MgPT0gMV0gLyAgblRbbmV3bmVzcyA9PSAxXSxcbiAgICAgbWFpbiA9ICdSZWFkcyBmcm9tIGxhYmVsZWQgUk5BJyxcbiAgICAgeGxhYiA9ICdNdXRhdGlvbiByYXRlJyxcbiAgICAgY29sID0gJ2RhcmtyZWQnKVxuaGlzdChUQ1tuZXduZXNzID09IDBdIC8gIG5UW25ld25lc3MgPT0gMF0sXG4gICAgIG1haW4gPSAnUmVhZHMgZnJvbSB1bmxhYmVsZWQgUk5BJyxcbiAgICAgeGxhYiA9ICdNdXRhdGlvbiByYXRlJyxcbiAgICAgY29sID0gJ2RhcmtncmF5JykiLCJhdHRyIjp7ImVkaXQiOnRydWUsImV2YWwiOnRydWV9fQ==
</script>
</div>
</div>
<p>Some observations from the above visualizations:</p>
<ol type="1">
<li>Some reads from unlabeled RNA have mutations. This is due to the non-zero background mutation rate.</li>
<li>Some reads from labeled RNA have no mutations. This is due to the incorporation rate being much less than 100%. Experimentally, this is due to the metabolic label having to compete with the regular nucleotide for incorporation into nascent RNA.</li>
</ol>
<p>These observations are what make analyzing NR-seq data challenging, and what will motivate analysis strategies discussed in other posts.</p>
</section>
<section id="rna-kinetics-and-nr-seq" class="level3">
<h3 class="anchored" data-anchor-id="rna-kinetics-and-nr-seq">RNA kinetics and NR-seq</h3>
<p>Different transcript isoforms can have very different properties. Of importance to an NR-seq experiment, different isoforms can differ in their nucleotide content and turnover kinetics. The latter’s impact is obvious, but my reason for bringing up the former may not be. In short, the rate at which a given isoform is degraded will determine its fraction new.</p>
<p>To see why turnover kinetics of RNA influences NR-seq data, consider the following model:</p>
<p>In it, RNA’s are synthesized at some rate ksyn, and degraded with a rate constant kdeg. kdeg is related to the average lifetime of a given RNA. When the rate at which a given species of RNA is degraded (kdeg * the amount of RNA that exists to be degraded) is equal to the rate at which it is synthesized (ksyn), that RNA is said to be at steady-state. A bit of algebra reveals that this occurs when the levels of RNA are equal to ratio of the synthesis and degradation rate constants:</p>
<p><span class="math display">\[
\begin{aligned}
\text{ksyn} &amp; = \text{kdeg}*[\text{RNA}]_{\text{ss}} \\
\frac{\text{ksyn}}{\text{kdeg}} &amp; = [\text{RNA}]_{\text{ss}}
\end{aligned}
\]</span></p>
<p>When you add metabolic label, you effectively create two species of RNA with distinct dynamics:</p>
<ol type="1">
<li>Old RNA that existed at the time of labeling. These can only degrade and are no longer synthesized. In this model, this means that they exponentially degrade with rate constant kdeg.</li>
<li>New RNA that is synthesized during labeling. These will slowly accumulate to steady-state levels as they are both synthesized and degraded.</li>
</ol>
<p>Assuming that the synthesis and degradation rate constants are constant throughout the label time, the dynamics of old and new RNA looks like:</p>
<div class="cell">
<div>
<div id="webr-4" class="exercise-cell">

</div>
<script type="webr-4-contents">
eyJjb2RlIjoiIyBRdWFydG8gbGl2ZSBjYW4gYmUgd2VpcmQgc29tZXRpbWVzXG4jIEJlc3QgdG8gZXhwbGljaXRseSBzZXQgYmFja2dyb3VuZCBhcyB3aGl0ZVxucGFyKGJnID0gXCJ3aGl0ZVwiKVxuXG4jIyMgU2ltdWxhdGlvbiBwYXJhbWV0ZXJzXG5rc3luIDwtIDEwICMgbW9sZWN1bGVzIHBlciBob3VyXG5rZGVnIDwtIDAuMiAjIDEgLyAgaG91clxudGVuZCA8LSAyNCAjIGhvdXJzXG50aW1lc3RlcCA8LSAwLjAxICMgaG91cnNcblxuIyMjIFNpbXVhdGUgZHluYW1pY3NcbnRpbWUgPC0gc2VxKGZyb20gPSAwLCB0byA9IHRlbmQsIGJ5ID0gdGltZXN0ZXApXG5Sc3MgPC0ga3N5biAvIGtkZWdcbk9sZF9SIDwtIFJzcyAqIGV4cCgta2RlZyp0aW1lKVxuTmV3X1IgPC0gUnNzICogKDEgLSBleHAoLWtkZWcqdGltZSkpXG5cbiMjIyBQbG90XG50aWJibGUoUk5BID0gYyhPbGRfUiwgTmV3X1IpLFxuICAgICAgIHBvcHVsYXRpb24gPSBmYWN0b3IocmVwKGMoXCJPbGRcIiwgXCJOZXdcIiksIGVhY2ggPSBsZW5ndGgoT2xkX1IpKSwgbGV2ZWxzID0gYyhcIk9sZFwiLCBcIk5ld1wiKSksXG4gICAgICAgdGltZSA9IHJlcCh0aW1lLCB0aW1lcyA9IDIpKSAlPiVcbiAgZ2dwbG90KGFlcyh4ID0gdGltZSwgeSA9IFJOQSwgY29sb3IgPSBwb3B1bGF0aW9uKSkgKyBcbiAgZ2VvbV9saW5lKCkgKyBcbiAgdGhlbWVfY2xhc3NpYygpICsgXG4gIHNjYWxlX2NvbG9yX21hbnVhbCh2YWx1ZXMgPSBjKCdkYXJrZ3JheScsICdkYXJrcmVkJykpICtcbiAgeGxhYihcIlRpbWUgc2luY2UgbGFiZWwgaW50cm9kdWN0aW9uXCIpICsgXG4gIHlsYWIoXCJbUk5BXVwiKSArXG4gIGdlb21faGxpbmUoeWludGVyY2VwdCA9IFJzcywgXG4gICAgICAgICAgICAgY29sb3IgPSAnY29yYWwnLCBcbiAgICAgICAgICAgICBsaW5ldHlwZSA9ICdkb3R0ZWQnLFxuICAgICAgICAgICAgIGxpbmV3aWR0aCA9IDEpIiwiYXR0ciI6eyJlZGl0Ijp0cnVlLCJldmFsIjp0cnVlfX0=
</script>
</div>
</div>
<p>Because the rate constants are unchanging, so is the total amount of RNA at any given time. Thus, the amount of Old RNA + New RNA = steady-state RNA level. Because of this, there is a simple relationship between the turnover kinetics of an RNA and the fraction new (abbreviated fn) for that RNA:</p>
<p><span class="math display">\[
\begin{aligned}
\text{fn} &amp;= \frac{[\text{New RNA}]}{[\text{RNA}]} \\
\text{fn} &amp;= \frac{[\text{RNA}]_{\text{ss}}\ast(1 - e^{-\text{kdeg}\ast \text{tl}})}{[\text{RNA}]_{\text{ss}}} \\
\text{fn} &amp;= 1 - e^{-\text{kdeg}\ast \text{tl}}
\end{aligned}
\]</span> where tl is the amount of time for which the cells were labeled. This simple relationship reveals some of the power of NR-seq. NR-seq provides information about RNA kinetics inaccessible to analyses of standard RNA-seq data (i.e., read counts). We can also use this relationship to modify our simulation and set the more biologically interpretable degradation rate constant rather than the fraction new:</p>
<div class="cell">
<div>
<div id="webr-5" class="exercise-cell">

</div>
<script type="webr-5-contents">
eyJjb2RlIjoiIyBRdWFydG8gbGl2ZSBjYW4gYmUgd2VpcmQgc29tZXRpbWVzXG4jIEJlc3QgdG8gZXhwbGljaXRseSBzZXQgYmFja2dyb3VuZCBhcyB3aGl0ZVxucGFyKGJnID0gXCJ3aGl0ZVwiKVxuXG4jIyMgU2ltdWxhdGlvbiBwYXJhbWV0ZXJzXG5yZWFkcyA8LSAxMDBcblVjb250IDwtIDAuMjVcbnJlYWRsZW4gPC0gMTUwXG5rZGVnIDwtIDAuMlxudGwgPC0gNFxucG5ldyA8LSAwLjA1XG5wb2xkIDwtIDAuMDAyXG5cblxuIyMjIG5UIFNpbXVsYXRpb25cbm5UIDwtIHJiaW5vbShyZWFkcywgc2l6ZSA9IHJlYWRsZW4sIHByb2IgPSBVY29udClcblxuXG4jIyMgTXV0YXRpb25hbCBjb250ZW50IHNpbXVsYXRpb25cbmZuIDwtIDEgLSBleHAoLWtkZWcqdGwpXG5uZXduZXNzIDwtIHJiaW5vbShyZWFkcywgc2l6ZSA9IDEsIHByb2IgPSBmbilcblRDIDwtIHJiaW5vbShyZWFkcywgc2l6ZSA9IG5ULCBwcm9iID0gcG5ldypuZXduZXNzICsgcG9sZClcblxuIyMjIFZpc3VhbGl6ZVxuaGlzdChUQyAvIG5ULFxuICAgICBtYWluID0gJ0FsbCByZWFkcycsXG4gICAgIGNvbCA9ICdjb3JhbCcpXG5oaXN0KFRDW25ld25lc3MgPT0gMV0gLyAgblRbbmV3bmVzcyA9PSAxXSxcbiAgICAgbWFpbiA9ICdSZWFkcyBmcm9tIGxhYmVsZWQgUk5BJyxcbiAgICAgeGxhYiA9ICdNdXRhdGlvbiByYXRlJyxcbiAgICAgY29sID0gJ2RhcmtyZWQnKVxuaGlzdChUQ1tuZXduZXNzID09IDBdIC8gIG5UW25ld25lc3MgPT0gMF0sXG4gICAgIG1haW4gPSAnUmVhZHMgZnJvbSB1bmxhYmVsZWQgUk5BJyxcbiAgICAgeGxhYiA9ICdNdXRhdGlvbiByYXRlJyxcbiAgICAgY29sID0gJ2RhcmtncmF5JykiLCJhdHRyIjp7ImVkaXQiOnRydWUsImV2YWwiOnRydWV9fQ==
</script>
</div>
</div>
</section>
<section id="simulating-multiple-transcripts" class="level3">
<h3 class="anchored" data-anchor-id="simulating-multiple-transcripts">Simulating multiple transcripts</h3>
<p>So far, we have focused on simulating data for a single transcript. In actuality, NR-seq is a high throughput method that provides information about all of the appreciably expressed transcripts in the cells from which you extracted RNA. Thus, we can set the number of transcripts we would like to simulate, and then draw kinetic parameters from a chosen distribution.</p>
<p>One thing we need to consider though is how the rate constants give rise to expected read counts for each transcript. We have discussed how the steady-state levels of a given RNA are a function of its synthesis and degradation rate constants, but is an RNA’s abundance related to its RNA-seq coverage?</p>
<p>The answer comes from realizing that RNA-seq is a measure of <strong>relative RNA abundance</strong>, not absolute abundance. That is, the number of reads you get from an RNA is a function of how abundant that RNA is, relative to all other RNAs in the sequenced pool. In addition, abundance in this setting is a function not only of the molecular abundance of an RNA (i.e., the number of molecules of that RNA present in the average cell), but also of the length of the RNA. This is because we are sequencing short fragments of an RNA, and thus the probability that we sequence a fragment from a given RNA depends on how many fragments in our pool come from that RNA. This is roughly <span class="math inline">\([\text{RNA}]_{\text{ss}} \ast \text{Length}\)</span>. In this simulation, we will make the simplifying assumption that all transcripts are the same length, and will thus only need to consider the relative steady-state abundances of each RNA. Thus, the parameters that we need to set are:</p>
<ol type="1">
<li>Those determining the rate constant distributions from which we sample transcript-specific ksyn’s and kdeg’s.</li>
<li>The total number of reads in our library. These will be randomly divided among the various simulated transcripts.</li>
<li>Those determining the U-content distribution from which we sapmle transcript-specific U-content’s.</li>
</ol>
<p>Deciding 1) and 3) is often best done through comparing candidate distributions to real data, and choosing parameters that make the simulated data look as close to the real data as possible. The result of this may look like:</p>
<div class="cell">
<div>
<div id="webr-6" class="exercise-cell">

</div>
<script type="webr-6-contents">
eyJjb2RlIjoiXG4jIyMgU2ltdWxhdGlvbiBwYXJhbWV0ZXJzXG5cbiMgTnVtYmVyIG9mIHRyYW5zY3JpcHRzIHRvIHNpbXVsYXRlXG5udCA8LSAxMDBcblxuIyBEaXN0cmlidXRpb24gcGFyYW1ldGVyc1xua2RlZ19sb2dtZWFuIDwtIC0xLjlcbmtkZWdfbG9nc2QgPC0gMC43XG5rc3luX2xvZ21lYW4gPC0gMi4zXG5rc3luX2xvZ3NkIDwtIDAuN1xuVWNvbnRfYWxwaGEgPC0gMjVcblVjb250X2JldGEgPC0gNzVcblxuIyBFeHBlcmltZW50YWwgcGFyYW1ldGVyc1xuc2VxZGVwdGggPC0gMjUwMDBcbnJlYWRsZW4gPC0gMTUwXG50bCA8LSA0XG5wbmV3IDwtIDAuMDVcbnBvbGQgPC0gMC4wMDJcblxuXG4jIyMgU2ltdWxhdGUgdHJhbnNjcmlwdCBwYXJhbWV0ZXJzXG5rZGVncyA8LSBybG5vcm0obnQsIGtkZWdfbG9nbWVhbiwga2RlZ19sb2dzZClcbmtzeW5zIDwtIHJsbm9ybShudCwga3N5bl9sb2dtZWFuLCBrc3luX2xvZ3NkKVxuUnNzIDwtIGtzeW5zIC8ga2RlZ3NcbnJlbF9hYnVuZGFuY2UgPC0gUnNzIC8gc3VtKFJzcylcbmZucyA8LSAxIC0gZXhwKC1rZGVncyp0bClcblxuVWNvbnRzIDwtIHJiZXRhKG50LCBVY29udF9hbHBoYSwgVWNvbnRfYmV0YSlcblxuIyMjIFNpbXVsYXRlIHJlYWQgY291bnRzIGZvciBlYWNoIHRyYW5zY3JpcHRcbnJlYWRzX3Blcl90IDwtIHJtdWx0aW5vbSgxLCBcbiAgICAgICAgICAgICAgICAgICAgICAgICBzaXplID0gc2VxZGVwdGgsXG4gICAgICAgICAgICAgICAgICAgICAgICAgcHJvYiA9IHJlbF9hYnVuZGFuY2UpWywxXVxuXG4jIyMgU2ltdWxhdGUgcmVhZC1zcGVjaWZpYyBkYXRhXG5cbm5ld25lc3MgPC0gcmJpbm9tKHNlcWRlcHRoLFxuICAgICAgICAgICAgICAgICAgc2l6ZSA9IDEsXG4gICAgICAgICAgICAgICAgICBwcm9iID0gcmVwKGZucywgdGltZXMgPSByZWFkc19wZXJfdCkpXG5cbm5UIDwtIHJiaW5vbShzZXFkZXB0aCxcbiAgICAgICAgICAgICBzaXplID0gcmVhZGxlbixcbiAgICAgICAgICAgICBwcm9iID0gcmVwKFVjb250cywgdGltZXMgPSByZWFkc19wZXJfdCkpXG5cblRDIDwtIHJiaW5vbShzZXFkZXB0aCxcbiAgICAgICAgICAgICBzaXplID0gblQsXG4gICAgICAgICAgICAgcHJvYiA9IHBuZXcgKiBuZXduZXNzICsgcG9sZClcblxuIyMjIENvbXBpbGUgZGF0YVxuXG5zaW1fZGYgPC0gdGliYmxlKFxuICB0cmFuc2NyaXB0ID0gcGFzdGUwKFwidHJhbnNjcmlwdFwiLCByZXAoMTpudCwgdGltZXMgPSByZWFkc19wZXJfdCkpLFxuICBUQyA9IFRDLFxuICBuVCA9IG5ULFxuICBuZXduZXNzID0gbmV3bmVzc1xuKVxuXG5wYXJfZGYgPC0gdGliYmxlKFxuICB0cmFuc2NyaXB0ID0gcGFzdGUwKFwidHJhbnNjcmlwdFwiLCAxOm50KSxcbiAga2RlZyA9IGtkZWdzLFxuICBrc3luID0ga3N5bnMsXG4gIGZuID0gZm5zLFxuICByZWFkcyA9IHJlYWRzX3Blcl90XG4pXG5cbiMjIyBWaXN1YWxpemVcbiMgUGFuZWwgMToga2RlZyBkaXN0cmlidXRpb25cbiMgUGFuZWwgMjoga3N5biBkaXN0cmlidXRpb25cbiMgUGFuZWwgMzogZm4gZGlzdHJpYnV0aW9uXG4jIFBhbmVsIDQ6IHJlYWQgY291bnQgZGlzdHJpYnV0aW9uXG4jIFBhbmVsIDU6IG5ldyByZWFkIG11dGF0aW9uIGRpc3RyaWJ1dGlvblxuIyBQYW5lbCA2OiBvbGQgcmVhZCBtdXRhdGlvbiBkaXN0cmlidXRpb25cbnAxIDwtIGdncGxvdChwYXJfZGYsXG4gICAgICAgICAgICAgYWVzKHggPSBrZGVnKSkgK1xuICBnZW9tX2hpc3RvZ3JhbSgpICtcbiAgdGhlbWVfY2xhc3NpYygpICsgXG4gIHhsYWIoXCJrZGVnJ3NcIikgKyBcbiAgeWxhYihcInRyYW5zY3JpcHQgY291bnRcIilcblxucDIgPC0gZ2dwbG90KHBhcl9kZixcbiAgICAgICAgICAgICBhZXMoeCA9IGtzeW4pKSArXG4gIGdlb21faGlzdG9ncmFtKCkgK1xuICB0aGVtZV9jbGFzc2ljKCkgKyBcbiAgeGxhYihcImtzeW4nc1wiKSArIFxuICB5bGFiKFwidHJhbnNjcmlwdCBjb3VudFwiKVxuXG5wMyA8LSBnZ3Bsb3QocGFyX2RmLFxuICAgICAgICAgICAgIGFlcyh4ID0gZm5zKSkgK1xuICBnZW9tX2hpc3RvZ3JhbSgpICtcbiAgdGhlbWVfY2xhc3NpYygpICsgXG4gIHhsYWIoXCJGcmFjdGlvbiBuZXdcIikgKyBcbiAgeWxhYihcInRyYW5zY3JpcHQgY291bnRcIilcblxucDQgPC0gZ2dwbG90KHBhcl9kZixcbiAgICAgICAgICAgICBhZXMoeCA9IGxvZzEwKHJlYWRzICsgMSkpKSArXG4gIGdlb21faGlzdG9ncmFtKCkgK1xuICB0aGVtZV9jbGFzc2ljKCkgKyBcbiAgeGxhYihcImxvZzEwKHJlYWRzICsgMSlcIikgKyBcbiAgeWxhYihcInRyYW5zY3JpcHQgY291bnRcIilcblxucDUgPC0gZ2dwbG90KHNpbV9kZiAlPiUgXG4gICAgICAgICAgICAgICBkcGx5cjo6ZmlsdGVyKG5UID4gMCAmIG5ld25lc3MgPT0gMCksXG4gICAgICAgICAgICAgYWVzKHggPSBUQyAvIG5UKSkgK1xuICBnZW9tX2hpc3RvZ3JhbShmaWxsID0gJ2RhcmtncmF5JykgK1xuICB0aGVtZV9jbGFzc2ljKCkgKyBcbiAgeGxhYihcIk11dGF0aW9uIHJhdGUgZm9yIG9sZCByZWFkc1wiKSArIFxuICB5bGFiKFwidHJhbnNjcmlwdCBjb3VudFwiKVxuXG5wNiA8LSBnZ3Bsb3Qoc2ltX2RmICU+JSBcbiAgICAgICAgICAgICAgIGRwbHlyOjpmaWx0ZXIoblQgPiAwICYgbmV3bmVzcyA9PSAxKSxcbiAgICAgICAgICAgICBhZXMoeCA9IFRDIC8gblQpKSArXG4gIGdlb21faGlzdG9ncmFtKGZpbGwgPSAnZGFya3JlZCcpICtcbiAgdGhlbWVfY2xhc3NpYygpICsgXG4gIHhsYWIoXCJNdXRhdGlvbiByYXRlIGZvciBuZXcgcmVhZHNcIikgKyBcbiAgeWxhYihcInRyYW5zY3JpcHQgY291bnRcIilcblxuXG5ncmlkLmFycmFuZ2UocDEsIHAyLCBwMywgcDQsIHA1LCBwNixcbiAgICAgICAgICAgICBucm93ID0gMyxcbiAgICAgICAgICAgICBuY29sID0gMikiLCJhdHRyIjp7ImVkaXQiOnRydWUsImV2YWwiOnRydWV9fQ==
</script>
</div>
</div>
</section>
</section>
<section id="summary" class="level2">
<h2 class="anchored" data-anchor-id="summary">Summary</h2>
<p>In this post, we introduced NR-seq, and built a simulation of NR-seq data to explore several aspects of NR-seq data. In the next post, we will discuss various strategies for analyzing this data, using the simulation we built here to test our strategies.</p>


<script type="webr-data">
eyJwYWNrYWdlcyI6eyJwa2dzIjpbImV2YWx1YXRlIiwia25pdHIiLCJodG1sdG9vbHMiLCJkcGx5ciIsImdncGxvdDIiLCJncmlkRXh0cmEiXSwicmVwb3MiOltdfSwib3B0aW9ucyI6eyJiYXNlVXJsIjoiaHR0cHM6Ly93ZWJyLnItd2FzbS5vcmcvdjAuNC4yLyJ9LCJyZW5kZXJfZGYiOiJkZWZhdWx0In0=
</script>
<script type="ojs-module-contents">
{"contents":[{"cellName":"webr-widget-6","inline":false,"source":"{\n  // Wait for output to be written to the DOM, then trigger widget rendering\n  await _webr_value_6;\n  if (window.HTMLWidgets) {\n    window.HTMLWidgets.staticRender();\n  }\n  if (window.PagedTableDoc) {\n    window.PagedTableDoc.initAll();\n  }\n}\n","methodName":"interpretQuiet"},{"cellName":"webr-6","inline":false,"source":"viewof _webr_editor_6 = {\n  const { WebRExerciseEditor, b64Decode } = window._exercise_ojs_runtime;\n  const scriptContent = document.querySelector(`script[type=\\\"webr-6-contents\\\"]`).textContent;\n  const block = JSON.parse(b64Decode(scriptContent));\n\n  const options = Object.assign({ id: `webr-6-contents` }, block.attr);\n  const editor = new WebRExerciseEditor(webROjs.webRPromise, block.code, options);\n\n  return editor.container;\n}\n_webr_value_6 = webROjs.process(_webr_editor_6, {});\n","methodName":"interpret"},{"cellName":"webr-widget-5","inline":false,"source":"{\n  // Wait for output to be written to the DOM, then trigger widget rendering\n  await _webr_value_5;\n  if (window.HTMLWidgets) {\n    window.HTMLWidgets.staticRender();\n  }\n  if (window.PagedTableDoc) {\n    window.PagedTableDoc.initAll();\n  }\n}\n","methodName":"interpretQuiet"},{"cellName":"webr-5","inline":false,"source":"viewof _webr_editor_5 = {\n  const { WebRExerciseEditor, b64Decode } = window._exercise_ojs_runtime;\n  const scriptContent = document.querySelector(`script[type=\\\"webr-5-contents\\\"]`).textContent;\n  const block = JSON.parse(b64Decode(scriptContent));\n\n  const options = Object.assign({ id: `webr-5-contents` }, block.attr);\n  const editor = new WebRExerciseEditor(webROjs.webRPromise, block.code, options);\n\n  return editor.container;\n}\n_webr_value_5 = webROjs.process(_webr_editor_5, {});\n","methodName":"interpret"},{"cellName":"webr-widget-4","inline":false,"source":"{\n  // Wait for output to be written to the DOM, then trigger widget rendering\n  await _webr_value_4;\n  if (window.HTMLWidgets) {\n    window.HTMLWidgets.staticRender();\n  }\n  if (window.PagedTableDoc) {\n    window.PagedTableDoc.initAll();\n  }\n}\n","methodName":"interpretQuiet"},{"cellName":"webr-4","inline":false,"source":"viewof _webr_editor_4 = {\n  const { WebRExerciseEditor, b64Decode } = window._exercise_ojs_runtime;\n  const scriptContent = document.querySelector(`script[type=\\\"webr-4-contents\\\"]`).textContent;\n  const block = JSON.parse(b64Decode(scriptContent));\n\n  const options = Object.assign({ id: `webr-4-contents` }, block.attr);\n  const editor = new WebRExerciseEditor(webROjs.webRPromise, block.code, options);\n\n  return editor.container;\n}\n_webr_value_4 = webROjs.process(_webr_editor_4, {});\n","methodName":"interpret"},{"cellName":"webr-widget-3","inline":false,"source":"{\n  // Wait for output to be written to the DOM, then trigger widget rendering\n  await _webr_value_3;\n  if (window.HTMLWidgets) {\n    window.HTMLWidgets.staticRender();\n  }\n  if (window.PagedTableDoc) {\n    window.PagedTableDoc.initAll();\n  }\n}\n","methodName":"interpretQuiet"},{"cellName":"webr-3","inline":false,"source":"viewof _webr_editor_3 = {\n  const { WebRExerciseEditor, b64Decode } = window._exercise_ojs_runtime;\n  const scriptContent = document.querySelector(`script[type=\\\"webr-3-contents\\\"]`).textContent;\n  const block = JSON.parse(b64Decode(scriptContent));\n\n  const options = Object.assign({ id: `webr-3-contents` }, block.attr);\n  const editor = new WebRExerciseEditor(webROjs.webRPromise, block.code, options);\n\n  return editor.container;\n}\n_webr_value_3 = webROjs.process(_webr_editor_3, {});\n","methodName":"interpret"},{"cellName":"webr-widget-2","inline":false,"source":"{\n  // Wait for output to be written to the DOM, then trigger widget rendering\n  await _webr_value_2;\n  if (window.HTMLWidgets) {\n    window.HTMLWidgets.staticRender();\n  }\n  if (window.PagedTableDoc) {\n    window.PagedTableDoc.initAll();\n  }\n}\n","methodName":"interpretQuiet"},{"cellName":"webr-2","inline":false,"source":"viewof _webr_editor_2 = {\n  const { WebRExerciseEditor, b64Decode } = window._exercise_ojs_runtime;\n  const scriptContent = document.querySelector(`script[type=\\\"webr-2-contents\\\"]`).textContent;\n  const block = JSON.parse(b64Decode(scriptContent));\n\n  const options = Object.assign({ id: `webr-2-contents` }, block.attr);\n  const editor = new WebRExerciseEditor(webROjs.webRPromise, block.code, options);\n\n  return editor.container;\n}\n_webr_value_2 = webROjs.process(_webr_editor_2, {});\n","methodName":"interpret"},{"cellName":"webr-widget-1","inline":false,"source":"{\n  // Wait for output to be written to the DOM, then trigger widget rendering\n  await _webr_value_1;\n  if (window.HTMLWidgets) {\n    window.HTMLWidgets.staticRender();\n  }\n  if (window.PagedTableDoc) {\n    window.PagedTableDoc.initAll();\n  }\n}\n","methodName":"interpretQuiet"},{"cellName":"webr-1","inline":false,"source":"viewof _webr_editor_1 = {\n  const { WebRExerciseEditor, b64Decode } = window._exercise_ojs_runtime;\n  const scriptContent = document.querySelector(`script[type=\\\"webr-1-contents\\\"]`).textContent;\n  const block = JSON.parse(b64Decode(scriptContent));\n\n  const options = Object.assign({ id: `webr-1-contents` }, block.attr);\n  const editor = new WebRExerciseEditor(webROjs.webRPromise, block.code, options);\n\n  return editor.container;\n}\n_webr_value_1 = webROjs.process(_webr_editor_1, {});\n","methodName":"interpret"},{"cellName":"webr-prelude","inline":false,"source":"webROjs = {\n  const { WebR, ChannelType } = window._exercise_ojs_runtime.WebR;\n  const {\n    WebREvaluator,\n    WebREnvironmentManager,\n    setupR,\n    b64Decode,\n    collapsePath\n  } = window._exercise_ojs_runtime;\n\n  const statusContainer = document.getElementById(\"exercise-loading-status\");\n  const indicatorContainer = document.getElementById(\"exercise-loading-indicator\");\n  indicatorContainer.classList.remove(\"d-none\");\n\n  let statusText = document.createElement(\"div\")\n  statusText.classList = \"exercise-loading-details\";\n  statusText = statusContainer.appendChild(statusText);\n  statusText.textContent = `Initialise`;\n\n  // Hoist indicator out from final slide when running under reveal\n  const revealStatus = document.querySelector(\".reveal .exercise-loading-indicator\");\n  if (revealStatus) {\n    revealStatus.remove();\n    document.querySelector(\".reveal > .slides\").appendChild(revealStatus);\n  }\n\n  // Make any reveal slides with live cells scrollable\n  document.querySelectorAll(\".reveal .exercise-cell\").forEach((el) => {\n    el.closest('section.slide').classList.add(\"scrollable\");\n  })\n\n  // webR supplemental data and options\n  const dataContent = document.querySelector(`script[type=\\\"webr-data\\\"]`).textContent;\n  const data = JSON.parse(b64Decode(dataContent));\n\n  // Grab list of resources to be downloaded\n  const filesContent = document.querySelector(`script[type=\\\"vfs-file\\\"]`).textContent;\n  const files = JSON.parse(b64Decode(filesContent));\n\n  // Use PostMessage channel for async communication\n  // We do not expect to take nested prompt input in Quarto Live cells\n  data.options.channelType = ChannelType.PostMessage;\n\n  // Initialise webR and setup for R code evaluation\n  let webRPromise = (async (webR) => {\n    statusText.textContent = `Downloading webR`;\n    await webR.init();\n\n    // Install provided list of packages\n    // Ensure webR default repo is included\n    data.packages.repos.push(\"https://repo.r-wasm.org\")\n    await data.packages.pkgs.map((pkg) => () => {\n      statusText.textContent = `Downloading package: ${pkg}`;\n      return webR.evalRVoid(`\n        webr::install(pkg, repos = repos)\n        library(pkg, character.only = TRUE)\n      `, { env: {\n        pkg: pkg,\n        repos: data.packages.repos,\n      }});\n    }).reduce((cur, next) => cur.then(next), Promise.resolve());\n\n    // Download and install resources\n    await files.map((file) => async () => {\n      const name = file.substring(file.lastIndexOf('/') + 1);\n      statusText.textContent = `Downloading resource: ${name}`;\n      const response = await fetch(file);\n      if (!response.ok) {\n        throw new Error(`Can't download \\`${file}\\`. Error ${response.status}: \"${response.statusText}\".`);\n      }\n      const data = await response.arrayBuffer();\n\n      // Store URLs in the cwd without any subdirectory structure\n      if (file.includes(\"://\")) {\n        file = name;\n      }\n\n      // Collapse higher directory structure\n      file = collapsePath(file);\n\n      // Create directory tree, ignoring \"directory exists\" VFS errors\n      const parts = file.split('/').slice(0, -1);\n      let path = '';\n      while (parts.length > 0) {\n        path += parts.shift() + '/';\n        try {\n          await webR.FS.mkdir(path);\n        } catch (e) {\n          if (!e.message.includes(\"FS error\")) {\n            throw e;\n          }\n        }\n      }\n\n      // Write this file to the VFS\n      return await webR.FS.writeFile(file, new Uint8Array(data));\n    }).reduce((cur, next) => cur.then(next), Promise.resolve());\n\n    statusText.textContent = `Installing webR shims`;\n    await webR.evalRVoid(`webr::shim_install()`);\n\n    statusText.textContent = `WebR environment setup`;\n    await setupR(webR, data);\n\n    statusText.remove();\n    if (statusContainer.children.length == 0) {\n      statusContainer.parentNode.remove();\n    }\n    return webR;\n  })(new WebR(data.options));\n\n  // Keep track of initial OJS block render\n  const renderedOjs = {};\n\n  const process = async (context, inputs) => {\n    const webR = await webRPromise;\n    const evaluator = new WebREvaluator(webR, context)\n    await evaluator.process(inputs);\n    return evaluator.container;\n  }\n\n  return {\n    process,\n    webRPromise,\n    renderedOjs,\n  };\n}\n","methodName":"interpretQuiet"}]}
</script>
<div id="exercise-loading-indicator" class="exercise-loading-indicator d-none d-flex align-items-center gap-2">
<div id="exercise-loading-status" class="d-flex gap-2">

</div>
<div class="spinner-grow spinner-grow-sm">

</div>
</div>
<script type="vfs-file">
W10=
</script>
</section>

</main> <!-- /main -->
<script type="ojs-module-contents">
eyJjb250ZW50cyI6W119
</script>
<script type="module">
if (window.location.protocol === "file:") { alert("The OJS runtime does not work with file:// URLs. Please use a web server to view this document."); }
window._ojs.paths.runtimeToDoc = "../../posts/post-with-code";
window._ojs.paths.runtimeToRoot = "../..";
window._ojs.paths.docToRoot = "../..";
window._ojs.selfContained = false;
window._ojs.runtime.interpretFromScriptTags();
</script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>